File: wakakusa/demo/keys.rb
# The keyboard as a set of chords, and the same handlers in the
# application's menu bar. A script presses one with `key:cmd+s` and
# picks one with `menu:Save`.
require "wakakusa"

class Keys
  def initialize
    @count = 0
    @saved = 0
    @last = "-"
    @pasted = "(nothing)"
  end

  def save
    @saved = @count
  end

  def clear
    @count = 0
    @saved = 0
  end

  def copy_count
    clipboard_set("count=#{@count}")
  end

  def paste
    @pasted = clipboard_get
  end

  def typed(key)
    @last = key
  end

  def view
    column(spacing: 8.0, padding: 12.0) {
      text "count: #{@count}  saved: #{@saved}"
      text "last key: #{@last}"
      text "pasted: #{@pasted}"
      row(spacing: 6.0) {
        button("+1") { @count += 1 }
        button("save") { save }
        button("copy") { copy_count }
        button("paste") { paste }
      }
    }
  end
end

app = Keys.new

menu_item("Count", "Save") { app.save }
menu_item("Count", "Clear") { app.clear }

shortcut("cmd+s") { app.save }
shortcut("cmd+shift+r") { app.clear }
shortcut("cmd+shift+c") { app.copy_count }
shortcut("cmd+shift+v") { app.paste }
on_key { |chord| app.typed(chord) }

run(app, title: "keys")

File: gomamochi/demo/keys.go
// The keyboard as a set of chords, and the same handlers in the
// application's menu bar. A script presses one with `key:cmd+s` and
// picks one with `menu:Save`.
package main

import (
	"fmt"

	. "github.com/i2y/yokan/gomamochi"
)

type Keys struct {
	count  int
	saved  int
	last   string
	pasted string
}

func (k *Keys) save() {
	k.saved = k.count
}

func (k *Keys) clear() {
	k.count = 0
	k.saved = 0
}

func (k *Keys) copyCount() {
	ClipboardSetText(fmt.Sprintf("count=%d", k.count))
}

func (k *Keys) paste() {
	k.pasted = ClipboardGetText()
}

func (k *Keys) typed(chord string) {
	k.last = chord
}

func (k *Keys) View() Element {
	return Column(
		Text(fmt.Sprintf("count: %d  saved: %d", k.count, k.saved)),
		Text("last key: "+k.last),
		Text("pasted: "+k.pasted),
		Row(
			Button("+1").OnClick(func() { k.count += 1 }),
			Button("save").OnClick(func() { k.save() }),
			Button("copy").OnClick(func() { k.copyCount() }),
			Button("paste").OnClick(func() { k.paste() }),
		).Spacing(6),
	).Spacing(8).Padding(12)
}

func main() {
	app := &Keys{last: "-", pasted: "(nothing)"}

	MenuItem("Count", "Save", func() { app.save() })
	MenuItem("Count", "Clear", func() { app.clear() })

	Shortcut("cmd+s", func() { app.save() })
	Shortcut("cmd+shift+r", func() { app.clear() })
	Shortcut("cmd+shift+c", func() { app.copyCount() })
	Shortcut("cmd+shift+v", func() { app.paste() })
	OnKey(func(chord string) { app.typed(chord) })

	Run(app, Title("keys"))
}
